### PR TITLE
Cleaner handling of step limit

### LINKED ISSUE #934: Reporting steps limit exceeded is awkward
Currently, when Alt-Ergo reaches the steps limit given using `--steps-bound`, the reporting is weird, e.g.:
```
[Error]; Fatal Error: Steps limit reached: 10
```
There is no reason for considering the reaching of the steps limit as a "fatal error" or even an error. It should instead report a regular `unknown` or `I don't know` result, as it does when the time limit is reached

Moreover, it would be better to comply to the more or less SMT compliant answer, where one gets extra information using `(get-info :reason-unknown)`

### LINKED ISSUE #935: feature wish: improve the reporting of the number of steps taken
Currently, to get the number of steps taken by Alt-Ergo (using the Dolmen frontend) one has to parse the SMT comment of the form
```
; File \".*\", line [0-9]+, characters [0-9]+-[0-9]+:.*(\\([0-9]+.?[0-9]*\\) steps)
```
which indeed comes before the answer `unsat` or `unknown`. There should be a better way to get this info, e.g. by issuing a command of the form
```
(get-info :steps-taken)
```
after the `(check-sat)` query, and get back an answer of a standardized form, e.g.
```
resource::steps = ...
```

(moreover, this should be documented)

Thanks!

### PR DESCRIPTION
**Pull request by Stevendeo**
This PR aims at cleanly stop when the step limit is reached instead of returning a Fatal Error

### CODE REVIEW

File: src/bin/js/worker_js.ml
Comment (hra687261) [inline src/bin/js/worker_js.ml]: Suggestion:
```
          | Errors.Run_error _ -> 
            returned_status := Worker_interface.Error "Run error"
```

**Review by hra687261 — APPROVED** [review]
LGTM modulo a couple of superficial suggestions

File: src/lib/reasoners/sat_solver_sig.ml
Comment (hra687261) [inline src/lib/reasoners/sat_solver_sig.ml]: `Step_limit`*

File: src/lib/util/util.mli
Comment (bclement-ocp) [inline src/lib/util/util.mli:32]: Those two have a very similar role. We could factor that out into a single `exception Limit_reached of resources_limit` and `type resources_limit = Time | Steps of int` to ensure uniform handling.

(I am fine with keeping the PR as is and making this a long-term cleanup issue instead)

**Review by bclement-ocp — APPROVED** [review]

**Review by bclement-ocp — CHANGES_REQUESTED** [review]
I approved but actually — please add a test with a very low step limit (e.g. 1) and `(get-info :reason-unknown)` to ensure that:

 - We properly return `unknown` when reaching the step lmit, and
 - We properly say that the reason is due to the step limit

File: src/lib/util/util.mli
Comment (Stevendeo) [reply src/lib/util/util.mli:32]: I'll do the change right now, I like this factorization

Comment (bclement-ocp) [reply src/lib/util/util.mli:32]: (Note that the type should be `resource_limit`, not `resources_limit`, because of English rules being different from French ones)

Comment (Stevendeo) [reply src/lib/util/util.mli:32]: Actually this change is a little trickier than I thought. Right now the Timeout exception is catched by the fronend while, in my version, the limit_step exception is catched by the solver. I will leave it as it until we do a whole rework of the exception handling in the solver.

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [inline src/bin/common/solving_loop.ml]: I this is raised by `Steps.set_steps_bound` the `try .. with ..` block should only include `Steps.set_steps_bound` here.

File: src/bin/js/worker_js.ml
Comment (bclement-ocp) [inline src/bin/js/worker_js.ml]: @hra687261's suggestion of removing the match here still applies.

File: src/lib/reasoners/fun_sat.ml
Comment (bclement-ocp) [inline src/lib/reasoners/fun_sat.ml]: Can you explain why the steps limit increases while pushing? We call: `E.fresh_name`, `save_guard_and_refs` (which only pushes things on a stack), `mk_gf` (which is just a dumb record constructor) and `ME.add`, none of which seem to increase the step count.

File: src/lib/reasoners/sat_solver_sig.ml
Comment (bclement-ocp) [reply src/lib/reasoners/sat_solver_sig.ml]: @hra687261's remark still applies.

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml]: Returning `env` here is not correct for satML. Because the underlying satml solver uses mutability, it is shared between `env` and `acc`; if we reach the steps limit here, `new_vars` (called through `create_guard`) can leave the satml solver in an inconsistent state with variables partially added. Plus, `env` will be out of sync with the content of the satml solver and weird things will follow.

I see two solutions here — either we make sure that `new_vars` returns a valid environment here, so that `acc` is correct in the exception handler and we can then pop a corresponding amount, or we temporarily disable the step limit while pushing. I'd prefer the second option (even though it can cause going slightly over the step limit) because the first one is opening up a can of worms where we *will* mess up and introduce subtle bugs in the future.

**Review by bclement-ocp — CHANGES_REQUESTED** [review]
I don't think we can do this this way (see also #946 )

For satML in particular we would need to make sure that whenever the "step limit reached" exception is raised it cannot leave the solver in an invalid intermediate state, which we will very probably mess up. I have made comments on that fact wrt `push` but I now realize that it applies more generally — the sat solver was written with the fact that it could get interrupted by an exception at any time and then re-used. Trying to make it do so is probably a recipe for many bugs ranging from crashes to soundness holes.

Fixing this properly would be a huge amount of extremely error-prone work and frankly I don't think that it is worth it. I'd be in favor of simply putting the solver in a "imited" state once the steps limit (or time limit, for that matter) is exhausted, and make it reject all commands except for diagnostic ones (e.g. `get-info`). This would fix #934 without an equivalent implicit requirement (you really shouldn't keep using the solver once the steps limit has been exceeded because its output can be complete garbage at that point).

**Comment by bclement-ocp** [comment]
> you really shouldn't keep using the solver once the steps limit has been exceeded because its output can be complete garbage at that point

Note that this really only applies to satML, not to the Tableaux solver, since the Tableaux solver is purely functional and won't get stuck in an intermediate invalid state. But satML is the default solver that 99% of users will use, so…

**Comment by bclement-ocp** [comment]
> Note that this really only applies to satML, not to the Tableaux solver, since the Tableaux solver is purely functional and won't get stuck in an intermediate invalid state. But satML is the default solver that 99% of users will use, so…

Well and even then only "pure" Tableaux, not Tableaux-CDCL, which has an internal CDCL solver that could potentially end up in an invalid state as well!

File: src/lib/reasoners/sat_solver_sig.ml
Comment (Stevendeo) [reply src/lib/reasoners/sat_solver_sig.ml]: This is just a copy of the previous ADT, which uses plural for steps. I may not be as fluent as you are in English, but my guess is "step limit" is okay to mean "this step is the limit" and "steps limit" is also ok to mean "this number of steps is the limit".

File: src/bin/common/solving_loop.ml
Comment (Stevendeo) [reply src/bin/common/solving_loop.ml]: If set_steps_bound fails, we do not want `State.set` to be called. It was written initially when I put "recoverable_error" instead of "fatal_error" in the next line.

File: src/lib/reasoners/sat_solver_sig.ml
Comment (bclement-ocp) [reply src/lib/reasoners/sat_solver_sig.ml]: It doesn't really matter (and I wouldn't have said anything if @hra687261 did not make a remark), but I'm pretty sure in this situation "steps limit" is incorrect, or at least not idiomatic (see for instance [here](https://english.stackexchange.com/questions/229406/how-to-correctly-write-chains-of-nouns-with-plurals-like-messages-list-and-ap)). It is a "resource limit", a "character limit", and a "step limit". This is different from French where we would say "la limite des pas", and "la limite du pas" would be wrong (I guess we could say "the limit of the steps" but it would be weird).

File: src/lib/reasoners/fun_sat.ml
Comment (Stevendeo) [reply src/lib/reasoners/fun_sat.ml]: I thought `fun_sat` and `satml_frontend` had the same behavior regarding the steps; I was wrong.

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [reply src/bin/common/solving_loop.ml]: In that case, the best way to write the code in modern OCaml is to use `match .. exception`:

```ocaml
match Steps.set_steps_bound i with
| () -> State.set steps_bound i st
| exception Invalid_argument _ ->
  fatal_error "..."
```

This helps controlling the scope of exception handling (I make the remark here because for standard exceptions such as `Not_found`, `Failure`, or `Invalid_argument` it is very easy to include an unrelated function in the `try .. with ..` block that could also raise that same exception and cause puzzling bugs).

File: src/lib/reasoners/satml_frontend.ml
Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: > Returning env here is not correct for satML. 

That's the point of the `TODO`, we need to actually cancel the push instead of just returning the env. But disabling the step limit is a better idea, I'll go for it.

Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: What I mean by "not correct" here is worse — it will cause the type of crashes that we see in #946 , see my toplevel comment.

File: src/lib/reasoners/sat_solver_sig.ml
Comment (Stevendeo) [reply src/lib/reasoners/sat_solver_sig.ml]: Just noted I defined the exception Step_limit_reached :sweat_smile:

**Comment by Stevendeo** [comment]
Just in case, I disactivated the steps exception at the Frontend level to make sure we never break any invariant in the future.

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml:1273]: Why not directly call `Steps.apply_without_step_limit` from inside `Satml_frontend.push`?

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml]: As I explained, this is probably incorrect (i.e. there is no guarantee that `env` is in a valid state), because there is no guarantee that the sat solver be in a consistent state when the `Step_limit` is reached.

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml]: Same as above

**Review by bclement-ocp — CHANGES_REQUESTED** [review]

File: src/lib/reasoners/satml_frontend.ml
Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml:1273]: > Just in case, I disactivated the steps exception at the Frontend level to make sure we never break any invariant in the future.

For this reason

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: For instance we could raise `Steps_limit` in the middle of the `Satml.theory_propagate` loop and end up with the queues in a weird state that may or may not be valid.

Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: I am adding a wrapper at the Frontend level that blocks the calls to the SAT (sorry it came just after the review request, I forgot to add it)

Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: It does not solve the issue where the number of steps is updated after failing once, which cannot be solved by putting the solver in a "limited" state without making it permanent.
I'm starting to think the fatal error was a good choice while there is no deep work on the solver to make sure environments are consistent even when failing in the middle of the execution of an instruction.

File: src/lib/reasoners/satml_frontend.ml
Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: By blocking the instructions at the frontend level, we ensure the environment cannot be used in a wrong way. Besides, we may want to display information from the invalid environment when the step limit is reached, so I would rather keep it like that.

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: > It does not solve the issue where the number of steps is updated after failing once, which cannot be solved by putting the solver in a "limited" state without making it permanent.

If the solver is in a "step limit exceeded" state, that shouldn't change if the step limit is updated, so that should be fine, no?

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: But if `Step_limit_reached` was raised, the `env` is potentially invalid to use. How do you avoid it being reused in this case?

File: src/lib/frontend/frontend.ml
Comment (bclement-ocp) [inline src/lib/frontend/frontend.ml]: These also need `wrap_f`

Comment (Stevendeo) [reply src/lib/frontend/frontend.ml]: I did not put the wrapper on these because I did not want to silently fail (fail = do nothing) on `push` and `pop`.

Comment (bclement-ocp) [reply src/lib/frontend/frontend.ml]: Ah, I did not catch this, but the wrapper should *not* silently fail, it should return an error.

If the underlying state of the SAT solver is invalid, there is no guarantee that `push` and `pop` are correct either, so this could possibly crash as well.

Comment (Stevendeo) [reply src/lib/frontend/frontend.ml]: Not now `push` can no longer fail with the disabled steps check

Comment (Stevendeo) [reply src/lib/frontend/frontend.ml]: Oh I see your point sorry, we may push a wrong environment; but is it really an issue?

Comment (bclement-ocp) [reply src/lib/frontend/frontend.ml]: The issue is not that `push` can fail. The issue is that if you call another function, say `assume`, and *that* function reaches the step limit, the environment becomes invalid (i.e. arbitrary internal invariants may be broken), and we shouldn't be calling `push` (or `pop`, for that matters) on an invalid environment.

Comment (bclement-ocp) [reply src/lib/frontend/frontend.ml]: > Oh I see your point sorry, we may push a wrong environment; but is it really an issue?

If a step limit is reached, there may or may not be some invariants in the environment that are broken. Without looking at the code, I don't know which invariants are broken, or how calling `push` or `pop` interacts with those broken invariants. It might be that this is perfectly fine, it might not be, but we won't know without tracking all the invariants that can potentially get broken, and even then if we can only call `push` and `pop` it is not very interesting, so we should just prevent that.

Comment (Stevendeo) [reply src/lib/frontend/frontend.ml]: The sat environment should keep track of its own inconsistency wrt the steps; with this, we could actually save invalid environments, update the step number and make sure we don't ever work with invalid environments. What do you think?

Comment (Stevendeo) [reply src/lib/frontend/frontend.ml]: Nevermind, it's pretty much what is already done. When the unknown_reason is Some (Step_limit _) we should not be able to do anything

Comment (bclement-ocp) [reply src/lib/frontend/frontend.ml]: I am not sure I understand why updating the step number seems to be such a huge issue. Once we have an unknown reason with `Step_limit` updating the step number is not going to change that unknown reason. What am I missing here?

Comment (Stevendeo) [reply src/lib/frontend/frontend.ml]: Now we use the Limit step as a reason not to keep using the environment, it is not an issue anymore.

File: src/lib/reasoners/fun_sat.ml
Comment (bclement-ocp) [inline src/lib/reasoners/fun_sat.ml]: I believe that this is still leaking a potentially invalid environment into the wild without *protecting* it with an unknown reason. This should raise `i_dont_know` and it should be caught in `handle_sat_exn` in the frontend to set the state there to `` `Unknown`` and following operations will be protected by `wrap_f`.

Edit: accidentally a word ("invalid")

Comment (bclement-ocp) [reply src/lib/reasoners/fun_sat.ml]: Sorry, this comment shouldn't be here — it should be in the corresponding position in `satml_frontend`

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml]: I believe that this is still leaking a potentially invalid environment into the wild without protecting it with an unknown reason. This should raise i_dont_know and it should be caught in handle_sat_exn in the frontend to set the state there to  `Unknown and following operations will be protected by wrap_f.

Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: If we raise `I_dont_know` we will answer `unknown` during an assertion; is that really what we want to do? I would rather update the environment with the correct `unknown_reason`

Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: > If we raise I_dont_know we will answer unknown during an assertion

If we raise `I_dont_know` I think it should be caught by the frontend which will update the unknown reason as needed, and only display `unknown` once the solving loop is done processing, but worth checking.

Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: The unknown reason is on the sat solver side

Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: Ah right — yeah returning the env with `Step_limit` unknown reason is what needs to happen then.

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [inline src/bin/common/solving_loop.ml]: Should this be taken care of by `DO.init`?

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [inline src/bin/common/solving_loop.ml:489]: Is it necessary to add this back?

File: src/lib/frontend/d_state_option.ml
Comment (bclement-ocp) [inline src/lib/frontend/d_state_option.ml]: Why `mstate`?

File: src/lib/frontend/d_state_option.mli
Comment (bclement-ocp) [inline src/lib/frontend/d_state_option.mli]: Should clarify "the toplevel Steps module" or some such — otherwise it is a bit confusing since it seems like this is interfacing with itself.

File: tests/smtlib/testfile-steps-bound.dolmen.smt2
Comment (bclement-ocp) [inline tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: I am not sure it is worth being able to define the `set-option` after stuff has been defined/pushed/etc. This is an option that should be supported at the top-level (i.e. before any constants/asserts) but we really don't need to support it past that — is there an use case here?

File: tests/smtlib/testfile-steps-bound.dolmen.expected
Comment (bclement-ocp) [inline tests/smtlib/testfile-steps-bound.dolmen.expected:3]: This is not a valid syntax. This should be `(:reason-unknown s)` where `s` is a valid s-expression, "Step limit: 3" is not a valid s-expression. It should be something similar to `(:reason-unknown (:step-limit 3))`.

**Review by bclement-ocp — CHANGES_REQUESTED** [review]
There are still some issues with the smt interface, see inline comments.

I wonder if we should call the option `steps-bound`. I know that this is the name of the CLI flag, but we use "limit" for the time limit and the Fourier-Motzkin cross-limit (not sure what that limits) and "limit" is closer to what it actually do. The other place we use "bound" is for the "age bound" that controls the behavior of matching and is not a resource limit. If we change the name, it should also be changed for the CLI options, so best to keep "steps-bound" for now and do the rename everywhere in a separate PR.

File: src/bin/common/solving_loop.ml
Comment (Stevendeo) [reply src/bin/common/solving_loop.ml:489]: This, like the previous one, are rebase artifacts; thanks for pointing them out.

File: src/bin/common/solving_loop.ml
Comment (Stevendeo) [reply src/bin/common/solving_loop.ml]: We don't even need to initialize this variable

File: tests/smtlib/testfile-steps-bound.dolmen.smt2
Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: Not a use case, but I did find a bug while implementing this when setting an option after a push, so I think we should keep the test as it is.

File: tests/smtlib/testfile-steps-bound.dolmen.expected
Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.expected:3]: https://github.com/OCamlPro/alt-ergo/issues/966

File: tests/smtlib/testfile-steps-bound.dolmen.smt2
Comment (bclement-ocp) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: I don't think we should have tests for features that we don't want to support, unless if the tests says that it should be an error. Otherwise, that is implicitly saying that we support the feature.

Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: Let's say that I want to support this feature then, the whole test is a use case (let's try something with a given limit, then with a greater limit).

Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: Besides, we don't want alt-ergo to throw random `assert false` even if the user writes non-sense in its smt file

Comment (bclement-ocp) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: You are making a false dichotomy in your second message. The options are not "support this use case" or "raise an assertion failure"; there is a mechanism for reporting proper errors to the user that can be used. If we currently raise an assertion failure for this test case, that is a bug, what we should do is display an appropriate error instead.

I don't really understand the point of setting a limit then a greater limit in the current infrastructure, since the solver will have to re-do everything it did the second time, so it seems to me you'd just be wasting time. If the internal solver infrastructure was able to support this and we could say "ignore the previously set limit and keep solving", why not, I'd be happy to throw that in. But currently, we can't; if the limit is reached, we have to start from scratch, which induces additional complexity (need to reset the solver, replay the assertion queue, make sure everything is properly synced, etc.) of the kind you just advocated for removing yesterday.

In any case: this is currently not supported (as evidenced by the fact it is neither documented nor working), and it would require committing to additional complexity to support properly given the current infrastructure (again, because we can't just re-use the solver). This doesn't sound like a decision that should be taken by introducing a test case in a PR; it sounds like a change that an issue should be created for and evaluated separately. 

Given that it is also a thing I have never needed when using solvers, nor heard anyone requesting — to my knowledge, resource limits are typically set once at the beginning and that's it — I would be interested if you have practical use cases for changing limits during execution (besides "just because"), but that discussion should be happening in a separate issue if it has to happen, not holding up the merge of this PR.

Comment (bclement-ocp) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: > Let's say that I want to support this feature then, the whole test is a use case (let's try something with a given limit, then with a greater limit).

Also note that this test uses `(reset)`, which will work and is easy to support. I am just asking for the `set-option :steps-bound` to be moved to the start of the file / immediately after the reset.

Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: > You are making a false dichotomy in your second message.
 
Sorry if I was not clear. I meant to say two separate things: that *I* did want this feature (because I'm tired of arguing for that, it is a *test*) ; and separately, that I didn't want `assert false` when setting an option between a push and a pop (which was a real issue when implementing, that this PR fixed). 

It is not even like push and pop had any effect on the matter, push and pop only affect the assertions. IMO, the simple fact that it triggered an assertion at some point may imply that we lack tests that push alt-ergo to its edge.

So, if you *really* want this line gone, can we at least agree on creating another test file that presses alt-ergo on strange cases? I would be fine keeping the line for this exact same reason.

Comment (bclement-ocp) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: > It is not even like push and pop had any effect on the matter, push and pop only affect the assertions. IMO, the simple fact that it triggered an assertion at some point may imply that we lack tests that push alt-ergo to its edge.

This is definitely 100% true, and somewhere where we have to improve! Not arguing with that at all.

> So, if you really want this line gone, can we at least agree on creating another test file that presses alt-ergo on strange cases? I would be fine keeping the line for this exact same reason.

(Note, I don't want that line gone — I want that line moved before the `push`)

I would be happy with having tests that explore corner cases and other oddities, that would be a good thing. What I am *not* happy with, however, is tests that test an unsupported feature (except to test that it fails with a proper error message, of course). Doing so enshrines the behavior, so that it effectively becomes supported: next time someone makes a change, the test will fail, and they will dutifully make sure that it still passes in the same way. That's the point of having tests, after all. I felt that I made clear that this was my stance, but I think this was partly in the meeting last yesterday and not in this discussion, so I apologize if there was confusion.

Now, here, given the current infrastructure and limitation regarding re-starting the solver after the step limit is reached, the feature set that we can safely commit to support in a coherent and satisfactory way is the use of `(set-option :step-bounds)` in `init` mode (i.e. before any `push`, `pop`, `assert`, etc.). Well, I guess that is not entirely true: we can support the use of `(set-option :step-bounds)` *as long as the step limit is not reached*, and then we can no longer allow increasing the step limit without additional infrastructure -- but that would make the validity of the command depend on internal Alt-Ergo details which is not good user experience.

Given that fact, and given the stance above, the tests should either be in that fragment, or if they are outside of that fragment they should either be errors or contain comments that say "hey we don't do the right thing here currently, it's OK if the test outcome change". But ideally such tests should be kept to a minimum, because it encourages a lot of churn in expected files which leads to paying less attention to changes in test outputs that can be wrong.

I am sorry to be arguing so much on this test, but hopefully my position makes sense here.

I think the options I would be happy with are:
 - Change the test so that the `(set-option)` are in init mode (i.e. before any `push`, `pop`, `declare`, `assert`, etc)
 - Keep the test as is, but change the code so that it *actually fails* instead of silently doing something (we already use an "are we in init mode" test for the "sat-solver" option, it can be reused from there)
 - Keep the test as is, but explicitly add a comment in the test file saying "this is unsupported but we want to make sure we don't crash".

Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: > Keep the test as is, but change the code so that it actually fails instead of silently doing something (we already use an "are we in init mode" test for the "sat-solver" option, it can be reused from there)

I am OK with this solution, but I think this belongs to another PR (that would slightly change the d_solver_options module).

Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: By the way, this test is expected to fail in a near future; the first assertion requires mores than 2 steps to be completed, but as it is the option *is* set at initialization because we do not treat SMT instructions imperatively. When we do, this test should fail.
I will add this fact to the test in case I am not assigned to the specific PR that will trigger the error.

Comment (bclement-ocp) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: Thank you, I am OK with this solution. I will approve the changes as soon as the info is added to the test case.

Can you point to the corresponding PR/issue related to making sure the test fails for tracking purposes, I am not sure which one it is.

Comment (Stevendeo) [reply tests/smtlib/testfile-steps-bound.dolmen.smt2:12]: Maybe something linked to #899 ? The issue is definitely linked to what we discussed on tuesday about the imperative execution of smtlib instructions.

**Review by bclement-ocp — APPROVED** [review]

**Comment by bclement-ocp** [comment]
(Waiting for the benches currently running on `next` to merge)

**Pull request merged**
